### PR TITLE
Make screen texture and depth texture work in Multiview

### PR DESCRIPTION
**Pull request by BastiaanOlij**
While populating the back buffers was already working for Multiview (in the Vulkan renderer), actually consuming it inside of the shaders was failing.

This PR adds the required logic for screen texture and depth textures to work in Multiview:
image

(the left square reads from screen texture and inverts the color, the right square reads from the depth buffer and shows a gradient).

Still to do:
- Implement back buffer copy in GLES3 and add multiview capabilities so this functionality works on GLES3 as well. After discussing with ClayJohn, opting for moving this until after we release Godot 4.0 as this is a relatively sizable job.

### CODE REVIEW

File: drivers/gles3/storage/material_storage.cpp [hunk 1]
```diff
@@ -1617,8 +1617,8 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["POINT_COORD"] = "gl_PointCoord";
 		actions.renames["INSTANCE_CUSTOM"] = "instance_custom";
 		actions.renames["SCREEN_UV"] = "screen_uv";
-		//actions.renames["SCREEN_TEXTURE"] = "color_buffer"; //Not implemented in 3D yet.
-		//actions.renames["DEPTH_TEXTURE"] = "depth_buffer"; // Not implemented in 3D yet.
+		actions.renames["SCREEN_TEXTURE"] = "screen_texture";
+		actions.renames["DEPTH_TEXTURE"] = "depth_buffer";
```
Comment (BastiaanOlij) [inline drivers/gles3/storage/material_storage.cpp:1621]: mental note for myself, add `actions.check_multiview_samplers`

**Comment by BastiaanOlij** [comment]
Ok, I've taken this as far as I can for Godot 4.0

On Vulkan everything seems to work just fine both on the mobile and clustered renderer. Both now read properly from screen texture and depth texture.

On OpenGLES I've made the compiler changes and fixed a few issues I ran into when testing with my test project. However I ran into the issue that backbuffer copying hadn't been implemented yet. While most of the old code from Godot 3 is still part of the renderer, it hasn't been updated for multiview. This we've decided to move till after Godot 4 is released.

File: drivers/gles3/shaders/scene.glsl [hunk 2]
```diff
@@ -725,8 +731,8 @@ void light_compute(vec3 N, vec3 L, vec3 V, float A, vec3 light_color, float atte
 #else
 		float D = D_GGX(cNdotH, alpha_ggx);
 		float G = V_GGX(cNdotL, cNdotV, alpha_ggx);
-#endif // LIGHT_ANISOTROPY_USED
-	   // F
+#endif // LIGHT_ANISOTROPY_USED \
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:734]: typo?

File: servers/rendering/renderer_rd/storage_rd/light_storage.cpp [hunk 3]
```diff
@@ -1467,7 +1467,6 @@ bool LightStorage::reflection_probe_instance_begin_render(RID p_instance, RID p_
 		fb.push_back(atlas->depth_buffer);
 		atlas->depth_fb = RD::get_singleton()->framebuffer_create(fb);
 
-		atlas->render_buffers->cleanup();
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/storage_rd/light_storage.cpp:1470]: is this related?

File: servers/rendering/shader_compiler.cpp [hunk 4]
```diff
@@ -1183,9 +1183,10 @@ String ShaderCompiler::_dump_node_code(const SL::Node *p_node, int p_level, Gene
 
 					code += "(";
 
-					// if normal roughness texture is used, we will add logic to automatically switch between
+					// if color backbuffer, depth backbuffer or normal roughness texture is used,
+					// we will add logic to automatically switch between
 					// sampler2D and sampler2D array and vec2 UV and vec3 UV.
-					bool normal_roughness_texture_used = false;
+					bool multiview_used = false;
```
Comment (clayjohn) [inline servers/rendering/shader_compiler.cpp:1189]:
```suggestion
					bool multiview_check_needed = false;
```

I don't think ``multiview_used`` is accurate for what the boolean is actually for

**Review by clayjohn — COMMENTED** [review]
Just some nitpicks, otherwise it looks fine!

File: drivers/gles3/shaders/scene.glsl (hunk 2 above)
Comment (BastiaanOlij) [reply drivers/gles3/shaders/scene.glsl:734]: Yeah I guess, but a weird one, I suspect it might have been my version of clang not making sense of the `// F` comment on the next line. I'm guessing that comment is bogus anyway

File: servers/rendering/renderer_rd/storage_rd/light_storage.cpp (hunk 3 above)
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/storage_rd/light_storage.cpp:1470]: No, just ran into it as I was debugging, it actually calls `cleanup` when it calls `configure` in the next line so it's double.

File: servers/rendering/shader_compiler.cpp (hunk 4 above)
Comment (BastiaanOlij) [reply servers/rendering/shader_compiler.cpp:1189]: good suggestion, I wasn't sure about the name either

File: drivers/gles3/shaders/scene.glsl (hunk 2 above)
Comment (BastiaanOlij) [reply drivers/gles3/shaders/scene.glsl]: Yeah turned out to be clang.. sigh

File: drivers/gles3/shaders/scene.glsl [hunk 5]
```diff
@@ -726,7 +732,7 @@ void light_compute(vec3 N, vec3 L, vec3 V, float A, vec3 light_color, float atte
 		float D = D_GGX(cNdotH, alpha_ggx);
 		float G = V_GGX(cNdotL, cNdotV, alpha_ggx);
 #endif // LIGHT_ANISOTROPY_USED
-	   // F
+		// F
```
Comment (akien-mga) [inline drivers/gles3/shaders/scene.glsl:735]: clang-format wants this to be reverted to what it was.

It's being buggy and trying to align the two comments, with the assumption that one tab == 4 spaces.
So it becomes:
```
#endif // LIGHT_ANISOTROPY_USED
             // F
```

Adding an empty line to separate them fixes it.

Comment (akien-mga) [reply drivers/gles3/shaders/scene.glsl:735]: Opened a bug report about this upstream:
- https://github.com/llvm/llvm-project/issues/60098

This case is just bad luck that `#endif // stuff` has the comment starting at column 7, and `<tab><tab>// F` has it starting at column 8. This looks like it's just off by one by mistake so clang-format decides to align them (despite us asking it not to with the `AlignTrailingComments: false` option).

Comment (BastiaanOlij) [reply drivers/gles3/shaders/scene.glsl]: Thats pretty annoying. I'm pretty sure we're going to have a repeat of this a few more times :)

File: drivers/gles3/storage/material_storage.cpp [hunk 6]
```diff
@@ -1640,8 +1640,8 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["NODE_POSITION_VIEW"] = "(model_matrix * scene_data.view_matrix)[3].xyz";
 
 		actions.renames["VIEW_INDEX"] = "ViewIndex";
-		actions.renames["VIEW_MONO_LEFT"] = "0";
-		actions.renames["VIEW_RIGHT"] = "1";
+		actions.renames["VIEW_MONO_LEFT"] = "uint(0)";
```
Comment (BastiaanOlij) [inline drivers/gles3/storage/material_storage.cpp:1643]: Before I forget, @clayjohn @dsnopek this was an interesting thing I ran into on OpenGL. When we're rendering in multiview `ViewIndex` is replaced by `gl_ViewID_OVR`, else it was replaced by `0`.

That is working fine in our Vulkan implementation. In OpenGL it was giving me problems because `gl_ViewID_OVR` is a `uint` and for some reason it's typing `0` and `1` as `int` and I got a type error on it. Hence the cast to `uint`.

This feels to me like a settings issue, it should be able to cast constants to either `uint` or `int` automatically as needed.

Comment (BastiaanOlij) [reply drivers/gles3/storage/material_storage.cpp:1643]: Note that I was testing with a shader in Godot that is actually using the ViewIndex variable.

Comment (clayjohn) [reply drivers/gles3/storage/material_storage.cpp:1643]: OpenGL doesn't like casting between int and uint, we are stuck being explicit everywhere unfortunately.

**Comment by dsnopek** [comment]
Sorry, missed this PR until just now!

The OpenGL renderer changes seem good to me! (Although, with the caveat that despite originally implementing the OpenGL multiview support, I am far from being an expert in rendering :-))

File: drivers/gles3/shaders/scene.glsl [hunk 7]
```diff
@@ -563,11 +563,17 @@ uniform highp samplerCubeShadow positional_shadow; // texunit:-4
 #endif // !defined(DISABLE_LIGHT_OMNI) || !defined(DISABLE_LIGHT_SPOT)
 
 #ifdef USE_MULTIVIEW
-uniform highp sampler2DArray depth_buffer; // texunit:-6
+uniform highp sampler2DArray depth_texture; // texunit:-6
```
Comment (clayjohn) [inline drivers/gles3/shaders/scene.glsl:566]: Will need a rebase, I renamed these to ``color_buffer`` and ``depth_buffer`` to match the other backends

File: drivers/gles3/storage/material_storage.cpp [hunk 8]
```diff
@@ -1617,8 +1617,8 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["POINT_COORD"] = "gl_PointCoord";
 		actions.renames["INSTANCE_CUSTOM"] = "instance_custom";
 		actions.renames["SCREEN_UV"] = "screen_uv";
-		//actions.renames["SCREEN_TEXTURE"] = "color_buffer"; //Not implemented in 3D yet.
-		//actions.renames["DEPTH_TEXTURE"] = "depth_buffer"; // Not implemented in 3D yet.
+		actions.renames["SCREEN_TEXTURE"] = "screen_texture";
+		actions.renames["DEPTH_TEXTURE"] = "depth_texture";
```
Comment (clayjohn) [inline drivers/gles3/storage/material_storage.cpp:1621]: Can be removed on rebase

File: servers/rendering/shader_compiler.cpp [hunk 9]
```diff
@@ -1287,21 +1288,27 @@ String ShaderCompiler::_dump_node_code(const SL::Node *p_node, int p_level, Gene
...
 									data_type_name = "multiviewSampler";
-									normal_roughness_texture_used = true;
+									multiview_check_needed = true;
 								} else {
 									data_type_name = ShaderLanguage::get_datatype_name(onode->arguments[i]->get_datatype());
 								}
 
 								code += data_type_name + "(" + node_code + ", " + sampler_name + ")";
+							} else if (RS::get_singleton()->is_low_end() && correct_texture_uniform) {
+								if (actions.check_multiview_samplers && (texture_uniform == "SCREEN_TEXTURE" || texture_uniform == "DEPTH_TEXTURE" || texture_uniform == "NORMAL_ROUGHNESS_TEXTURE")) {
```
Comment (clayjohn) [inline servers/rendering/shader_compiler.cpp:1300]: Now need to use ``is_screen_texture`` and ``is_depth_texture``

File: servers/rendering/shader_compiler.cpp [hunk 10]
```diff
@@ -1287,21 +1288,27 @@ String ShaderCompiler::_dump_node_code(const SL::Node *p_node, int p_level, Gene
 								}
 
 								String data_type_name = "";
-								if (texture_uniform == "NORMAL_ROUGHNESS_TEXTURE") {
+								if (actions.check_multiview_samplers && (is_screen_texture || texture_uniform == "DEPTH_TEXTURE" || texture_uniform == "NORMAL_ROUGHNESS_TEXTURE")) {
```
Comment (clayjohn) [inline servers/rendering/shader_compiler.cpp:1291]: will need to add an ``is_depth_texture`` and use it

**Review by clayjohn — COMMENTED** [review]
Just some comments to help out with the rebase

File: drivers/gles3/shaders/scene.glsl (hunk 7 above)
Comment (BastiaanOlij) [reply drivers/gles3/shaders/scene.glsl:566]: yeah I renamed to texture originally because we had that hardcoded in the compiler before all these changes.

**Comment by BastiaanOlij** [comment]
@clayjohn ok made all the changes and they work fine on Vulkan both mobile and clustered.

I'm having all sorts of issues with GLES though that I don't think are related to these changes but the original changes, for instance sky renderer doesn't seem to compile.

This is a little test project I've been using that you can run to also test stereo:

[Test stereo render.zip](https://github.com/godotengine/godot/files/10478532/Test.stereo.render.zip)

File: servers/rendering/shader_compiler.cpp [hunk 11]
```diff
@@ -1220,7 +1221,7 @@ String ShaderCompiler::_dump_node_code(const SL::Node *p_node, int p_level, Gene
 						}
 
 						String node_code = _dump_node_code(onode->arguments[i], p_level, r_gen_code, p_actions, p_default_actions, p_assigning);
-						if (!RS::get_singleton()->is_low_end() && is_texture_func && i == 1) {
```
Comment (clayjohn) [inline servers/rendering/shader_compiler.cpp:1223]: This is what is causing the issue on the OpenGL renderer. Removing this low end check makes it so that the shader compiler replaces the sampler name with ``sampler2D(sampler_name, sampler[])`` which we explicitly do not want to do in the openGL renderer as we use combined texture samplers there. 

For low end we need to add another condition to set ``multiview_check_needed``. I would undo this change here and then restructure the branch as follows:
```
if (!RS::get_singleton()->is_low_end() && is_texture_func && i == 1) {
	...
} else if (is_texture_func && i == 1) {
	multiview_check_needed = ... // detect if multiview_check_needed
	code += node_code;
} else if (multiview_check_needed && i == 2) {
	// UV coordinate after using color, depth or normal roughness texture.
	code += "multiview_uv(" + node_code + ".xy)";
} else {
	code += node_code;
}
```

Comment (BastiaanOlij) [reply servers/rendering/shader_compiler.cpp:1223]: Ah, that makes sense, it was getting confused with the merge on those lines and thought you changed it in another direction. ok will fix that up

Comment (BastiaanOlij) [reply servers/rendering/shader_compiler.cpp:1223]: I'm double checking `is_low_end` as it has little cost and properly writes out the scenario where we're checking for multiview for OpenGl, this in case someone ever changes or adds to the ifs checked and breaks the system :)

File: servers/rendering/shader_compiler.cpp [hunk 12]
```diff
@@ -1292,12 +1296,48 @@ String ShaderCompiler::_dump_node_code(const SL::Node *p_node, int p_level, Gene
...
+									texture_uniform = varnode->name;
+									correct_texture_uniform = true;
+								} break;
+								case SL::Node::TYPE_ARRAY: {
+									const SL::ArrayNode *anode = static_cast<const SL::ArrayNode *>(onode->arguments[i]);
+									texture_uniform = anode->name;
+									correct_texture_uniform = true;
+								} break;
+								default:
+									break;
```
Comment (clayjohn) [inline servers/rendering/shader_compiler.cpp:1318]: Hmmm, I guess this is all duplicated this way.

Sorry to nitpick again. But it looks like it would be better to instead introduce the low-end/heigh-end split at line 1244 (``if (correct_texture_uniform) {``)

So like:
```
if (correct_texture_uniform && !RS::get_singleton()->is_low_end()) {
	// Same as it is now
} else if (correct_texture_uniform && RS::get_singleton()->is_low_end()) {
	if (shader->uniforms.has(texture_uniform)) {
		const ShaderLanguage::ShaderNode::Uniform &u = shader->uniforms[texture_uniform];
		if (u.hint == ShaderLanguage::ShaderNode::Uniform::HINT_SCREEN_TEXTURE) {
			multiview_uv_needed = true;
		} else if (u.hint == ShaderLanguage::ShaderNode::Uniform::HINT_DEPTH_TEXTURE) {
			multiview_uv_needed = true;
		} else if (u.hint == ShaderLanguage::ShaderNode::Uniform::HINT_NORMAL_ROUGHNESS_TEXTURE) {
			multiview_uv_needed = true;
		}
	}
	code += node_code;
} else {
	// Same as it is now
}
```
And 1224 would go from:
```
if (!RS::get_singleton()->is_low_end() && is_texture_func && i == 1) {
```
to: 
```
if (is_texture_func && i == 1) {
```
As you already had it in the last iteration

Comment (BastiaanOlij) [reply servers/rendering/shader_compiler.cpp:1318]: Geeeeeeez.... :)

Just kidding, I was tossing a coin on that one when working on this.

**Review by clayjohn — APPROVED** [review]
Looks good! Sorry again for the nitpicking.

**Comment by akien-mga** [comment]
Thanks!

**Comment by BastiaanOlij** [comment]
> Looks good! Sorry again for the nitpicking.

Nah dude, the nitpicking is good, makes for a better solution :)
